examples/data-frames.dx

```diff
@@ -284,15 +235,62 @@ def crossJoin {n m f f'} (t:n=>{&...f}) (t':m=>{&...f'}) : (n & m)=>{...f & ...f
 -- It should be sufficient to do the join on a single column. One can always
 -- restructure the table so that the join happens on a tuple of values from
 -- the flattened columns.
-def leftJoin = todo
--- def leftJoin (c:Label) (l:n=>{@c: a & ...f}) (r:m=>{@c: a & ...f'})
---       : n=>{extra: Maybe {...f'} & ...f} = todo
+
+def findInUnordered {a n} [Eq a] (xs:n=>a) (v:a) : Maybe n =
+  -- Wasteful, but at least the parallelism is exposed.
+  (AsList num_found found_table) = argFilter (\x. x == v) xs
+  case num_found == 0 of
+    True -> Nothing
+    False -> Just found_table.(unsafeFromOrdinal _ 0)
```

Comment:
You could also do this:
```
data First a = MkFirst (Maybe a)

def FirstMonoid (a:Type) : Monoid (First a) =
  named-instance result : Monoid (First a)
    mempty = MkFirst Nothing
    mappend = \x y. case x of
      Just xv -> x
      Nothing -> y
  result

def findInUnordered {a n} [Eq a] (xs:n=>a) (v:a) : Maybe n =
  MkFirst ans = withAccum \ref.
    for_ i. case xs.i == v of
      True  -> ref += MkFirst (Just n)
      False -> ()
  ans
```

Reply:
Nice, yes I guess that'd be a bit more efficient.  I wonder if someday we can have some sort of `Lattice` typeclass and corresponding effect which lets the compiler know about short-circuiting completions.

Reply:
That would be pretty neat! We could have a flavor of Accum that works over a lattice and short-circuits (when the computation is otherwise pure?)!

Reply:
Yes, I think that'd be a good showcase of Dex's effect-aware compilation.  I've come across a couple of other examples that could in principle be sped up using a parallel short-circuit:
 - The `any` and `all` functions.
 - The `Ord` instance for lists which returns at the location of first difference